### PR TITLE
Support basic auth from URI in websockets

### PR DESCRIPTION
**Pull request by willhbr**
Just pipe the `.user` and `.password` fields from `URI` into the correct headers on the initial websocket request. Code stolen from `HTTP#basic_auth`.

### CODE REVIEW

**Comment by willhbr** [comment]
I'm not sure what to make of the MacOS CI failure - is this a known issue on main or did I somehow _only_ break MacOS?

File: src/http/web_socket/protocol.cr [hunk 1]
```diff
@@ -320,6 +320,9 @@ class HTTP::WebSocket::Protocol
 
     if (host = uri.hostname) && (path = uri.request_target)
       tls = uri.scheme.in?("https", "wss")
+      if (user = uri.user) && (password = uri.password)
+        headers["Authorization"] = "Basic #{Base64.strict_encode("#{user}:#{password}")}"
```
Comment (straight-shoota) [inline src/http/web_socket/protocol.cr:324]: This probably should not override an existing `Authorization` header. WDYT?
```suggestion
        headers["Authorization"] ||= "Basic #{Base64.strict_encode("#{user}:#{password}")}"
```

Comment (willhbr) [reply src/http/web_socket/protocol.cr:324]: I'd be happy with it either way - I can see reasonable arguments on either side. To be honest, throwing an exception is probably the safest (but maybe annoying in practice).

Not clobbering a custom header makes sense - if I was explicitly setting the header I can imagine being very frustrated at it being overridden, and probably less frustrated at the inverse.

**Comment by willhbr** [comment]
Any chance this can sneak in for 1.1?

**Comment by straight-shoota** [comment]
No way. We've been in the feature-freeze period for 1.1 before this PR was even created.

**Review by jhass — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

**Comment by willhbr** [comment]
Thanks! 🙂

**Pull request merged**